### PR TITLE
Relevance classification v2

### PR DESCRIPTION
**Pull request by georgerichardson**
Think this is all working ok. Relevance accuracy should be a lot better now.

### CODE REVIEW

File: source/python/idetect/nlp_models/base_model.py [hunk 1]
```diff
@@ -1,11 +1,18 @@
...
 import pandas as pd
 import requests
+import spacy
 from sklearn.externals import joblib
+from sklearn.base import TransformerMixin, BaseEstimator
+from scipy import sparse
+from gensim import matutils, models
+from gensim.sklearn_integration.sklearn_wrapper_gensim_lsimodel import SklLsiModel
 
+from idetect.geotagger import strip_accents, compare_strings, strip_words, common_names, LocationType, subdivision_country_code, match_country_name, city_subdivision_country
```
Comment (simonb83) [inline source/python/idetect/nlp_models/base_model.py:16]: I think there are a few unused imports, such as `spacy` and all the stuff from `idetect.geotagger`

File: source/python/idetect/nlp_models/relevance.py [hunk 2]
```diff
@@ -1,204 +1,235 @@
...
+    pos_tags : bool, required
+        Whether to tag words with their part of speech labels.
+    lemmatize : bool, required
+        Whether to lemmatize tokens.
+    stop_words : book, required
+        Whether to remove stop words.
+    """
+
+    def cleanup(self, text):
+        """
```
Comment (simonb83) [inline source/python/idetect/nlp_models/relevance.py:48]: You can import `cleanup` from `idetect.model` and that way we have a single function definition

**Review by simonb83 — APPROVED** [review]

File: source/python/idetect/nlp_models/base_model.py [hunk 3]
```diff
@@ -79,3 +86,61 @@ def predict(self, text):
...
+    """Gensim's Lsi model with sklearn wrapper, modified to handle sparse matrices
+    for both fit and transform. Makes the class compatible with sklearn's Tfidf and
+    Count vectorizers.
+    """
+
+    def sparse_2_tupes(self, sparse):
+        """Converts sparse matrix into manageable tuple format."""
+        for t in t_skltfidf:
+            cx = t.tocoo()
+            tups = []
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/base_model.py:101]: I think you can just say
```
tups = zip(cx.col, cx.data)
```
and skip the loop/append.

File: source/python/idetect/nlp_models/base_model.py [hunk 4]
```diff
@@ -79,3 +86,61 @@ def predict(self, text):
...
+        if sparse.issparse(docs):
+            docs = matutils.Sparse2Corpus(docs, documents_columns=False)
+        X = [[] for i in range(0, len(docs))];
+        for k,v in enumerate(docs):
+            doc_topics = self.gensim_model[v]
+            probs_docs = list(map(lambda x: x[1], doc_topics))
+            # Everything should be equal in length
+            if len(probs_docs) != self.num_topics:
+                probs_docs.extend([1e-12]*(self.num_topics - len(probs_docs)))
+            X[k] = probs_docs
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/base_model.py:144]: k is just a counter, right? Could this be simplified to 
```
X = []
for v in docs:
    ...
    X.append(probs_docs)
```

File: source/python/run_classifier.py [hunk 5]
```diff
@@ -2,10 +2,14 @@
 import sys
 
 from sqlalchemy import create_engine
-
+import re
```
Comment (WanderingStar) [inline source/python/run_classifier.py:5]: Are these added here because of the unpickling requirements?

File: source/python/idetect/nlp_models/base_model.py [hunk 6]
```diff
@@ -79,3 +86,61 @@ def predict(self, text):
...
+        # docs = check(docs)
+        if sparse.issparse(docs):
+            docs = matutils.Sparse2Corpus(docs, documents_columns=False)
+        X = [[] for i in range(0, len(docs))];
+        for k,v in enumerate(docs):
+            doc_topics = self.gensim_model[v]
+            probs_docs = list(map(lambda x: x[1], doc_topics))
+            # Everything should be equal in length
+            if len(probs_docs) != self.num_topics:
+                probs_docs.extend([1e-12]*(self.num_topics - len(probs_docs)))
```
Comment (WanderingStar) [inline source/python/idetect/nlp_models/base_model.py:143]: Why `1e-12` in particular? Just to get a very small positive number? Or because that value is special?
https://stackoverflow.com/questions/38477908/smallest-positive-float64-number

**Review by WanderingStar — APPROVED** [review]
Looks fine to me, generally. Reducing the extraneous inputs (if they are actually extraneous) might reduce the memory footprint of this and make it possible to run more of them on the Linode.
